File: day-one.cpp
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

/*
==============================
        IMPLEMENT
==============================
*/

ListNode* removeNthFromEnd(ListNode* head, int n) {
    // TODO
    return head;
}

ListNode* reverseList(ListNode* head) {
    // TODO
    return head;
}

ListNode* middleNode(ListNode* head) {
    // TODO
    return head;
}

ListNode* deleteMiddle(ListNode* head) {
    // TODO
    return head;
}

bool isPalindrome(ListNode* head) {
    // TODO
    return false;
}

/*
==============================
        HELPERS
==============================
*/

ListNode* build(const vector<int>& v) {
    if (v.empty()) return NULL;
    ListNode* head = new ListNode(v[0]);
    ListNode* cur = head;
    for (int i = 1; i < v.size(); i++) {
        cur->next = new ListNode(v[i]);
        cur = cur->next;
    }
    return head;
}

vector<int> toVec(ListNode* head) {
    vector<int> res;
    while (head) {
        res.push_back(head->val);
        head = head->next;
    }
    return res;
}

bool hasCycle(ListNode* head) {
    ListNode *slow = head, *fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) return true;
    }
    return false;
}

void assertSafe(ListNode* head, int expected_len) {
    assert(!hasCycle(head));
    assert((int)toVec(head).size() == expected_len);
}

void assertVec(const vector<int>& got, const vector<int>& exp) {
    if (got != exp) {
        cout << "\nExpected: ";
        for (auto x : exp) cout << x << " ";
        cout << "\nGot: ";
        for (auto x : got) cout << x << " ";
        cout << endl;
        assert(false);
    }
}

/*
==============================
            TESTS
==============================
*/

void test_removeNth() {
    auto ans1 = removeNthFromEnd(build({10}), 1);
    assertVec(toVec(ans1), {});
    assertSafe(ans1, 0);

    auto ans2 = removeNthFromEnd(build({1,2,3}), 3);
    assertVec(toVec(ans2), {2,3});
    assertSafe(ans2, 2);

    auto ans3 = removeNthFromEnd(build({1,2,3}), 1);
    assertVec(toVec(ans3), {1,2});
    assertSafe(ans3, 2);

    vector<int> v(40);
    iota(v.begin(), v.end(), 1);
    auto ans4 = removeNthFromEnd(build(v), 17);
    v.erase(v.begin() + (40-17));
    assertVec(toVec(ans4), v);
    assertSafe(ans4, v.size());
}

void test_reverse() {
    auto a = reverseList(NULL);
    assertVec(toVec(a), {});
    assertSafe(a, 0);

    auto b = reverseList(build({1}));
    assertVec(toVec(b), {1});
    assertSafe(b, 1);

    auto c = reverseList(build({1,2,3,4}));
    assertVec(toVec(c), {4,3,2,1});
    assertSafe(c, 4);

    // repeat reverse to catch corruption
    auto d = reverseList(c);
    assertVec(toVec(d), {1,2,3,4});
    assertSafe(d, 4);
}

void test_middle() {
    assert(middleNode(build({1}))->val == 1);
    assert(middleNode(build({1,2}))->val == 2);
    assert(middleNode(build({1,2,3,4,5,6}))->val == 4);
}

void test_deleteMiddle() {
    auto a = deleteMiddle(build({1}));
    assertVec(toVec(a), {});
    assertSafe(a, 0);

    auto b = deleteMiddle(build({1,2,3,4,5}));
    assertVec(toVec(b), {1,2,4,5});
    assertSafe(b, 4);
}

void test_palindrome() {
    auto a = build({1,2,3,2,1});
    assert(isPalindrome(a) == true);
    assertSafe(a, 5);  // ensure not destroyed

    auto b = build({1,2,3,4});
    assert(isPalindrome(b) == false);
    assertSafe(b, 4);
}

int main() {
    test_removeNth();
    test_reverse();
    test_middle();
    test_deleteMiddle();
    test_palindrome();

    cout << "ALL TESTS PASSED 🚀" << endl;
}

File: day-one.py
from typing import Optional, List


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


"""
==============================
        IMPLEMENT
==============================
"""

def removeNthFromEnd(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    # TODO
    return head


def reverseList(head: Optional[ListNode]) -> Optional[ListNode]:
    # TODO
    return head


def middleNode(head: Optional[ListNode]) -> Optional[ListNode]:
    # TODO
    return head


def deleteMiddle(head: Optional[ListNode]) -> Optional[ListNode]:
    # TODO
    return head


def isPalindrome(head: Optional[ListNode]) -> bool:
    # TODO
    return False


"""
==============================
        HELPERS
==============================
"""

def build(arr: List[int]) -> Optional[ListNode]:
    if not arr:
        return None
    head = ListNode(arr[0])
    cur = head
    for x in arr[1:]:
        cur.next = ListNode(x)
        cur = cur.next
    return head


def to_list(head: Optional[ListNode]) -> List[int]:
    res = []
    while head:
        res.append(head.val)
        head = head.next
    return res


def has_cycle(head: Optional[ListNode]) -> bool:
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False


def assert_safe(head: Optional[ListNode], expected_len: int):
    assert not has_cycle(head)
    assert len(to_list(head)) == expected_len


"""
==============================
            TESTS
==============================
"""

def test_removeNth():
    a = removeNthFromEnd(build([10]), 1)
    assert to_list(a) == []
    assert_safe(a, 0)

    b = removeNthFromEnd(build([1,2,3]), 3)
    assert to_list(b) == [2,3]
    assert_safe(b, 2)

    v = list(range(1, 41))
    exp = v.copy()
    exp.pop(len(v) - 17)
    c = removeNthFromEnd(build(v), 17)
    assert to_list(c) == exp
    assert_safe(c, len(exp))


def test_reverse():
    a = reverseList(None)
    assert to_list(a) == []
    assert_safe(a, 0)

    b = reverseList(build([1,2,3,4]))
    assert to_list(b) == [4,3,2,1]
    assert_safe(b, 4)

    # reverse again
    c = reverseList(b)
    assert to_list(c) == [1,2,3,4]
    assert_safe(c, 4)


def test_middle():
    assert middleNode(build([1])).val == 1
    assert middleNode(build([1,2])).val == 2
    assert middleNode(build([1,2,3,4,5,6])).val == 4


def test_deleteMiddle():
    a = deleteMiddle(build([1]))
    assert to_list(a) == []
    assert_safe(a, 0)

    b = deleteMiddle(build([1,2,3,4,5]))
    assert to_list(b) == [1,2,4,5]
    assert_safe(b, 4)


def test_palindrome():
    a = build([1,2,3,2,1])
    assert isPalindrome(a) is True
    assert_safe(a, 5)

    b = build([1,2,3,4])
    assert isPalindrome(b) is False
    assert_safe(b, 4)


if __name__ == "__main__":
    test_removeNth()
    test_reverse()
    test_middle()
    test_deleteMiddle()
    test_palindrome()

    print("ALL TESTS PASSED 🚀")
